manager/db.go

```diff
@@ -60,128 +66,249 @@ func (m *Manager) lookupCatalog(environmentId, name string) (model.Catalog, erro
 }
 
 func (m *Manager) updateDb(catalog model.Catalog, templates []model.Template, newCommit string) error {
-	tx := m.db.Begin()
+	// Pick out db Catalog ID
+	catalogQuery := `
+	SELECT *
+	FROM catalog
+	WHERE name = ? AND environment_id = ?`
 
-	if err := tx.Where(&model.CatalogModel{
-		Catalog: model.Catalog{
-			Name:          catalog.Name,
-			EnvironmentId: catalog.EnvironmentId,
-		},
-	}).Delete(&model.CatalogModel{}).Error; err != nil {
-		tx.Rollback()
-		return err
+	dbCatalog := model.CatalogModel{}
+	m.db.Raw(catalogQuery, catalog.Name, catalog.EnvironmentId).Find(&dbCatalog)
+
+	// Pick out db Template IDs
+	templateQuery := `
+	SELECT *
+	FROM catalog_template
+	WHERE catalog_id = ? AND environment_id = ?`
+	dbTemplates := []model.TemplateModel{}
+	m.db.Raw(templateQuery, dbCatalog.ID, catalog.EnvironmentId).Find(&dbTemplates)
+
+	dbTemplateIds := []uint{}
+	dbTemplateCommits := make(map[string]bool)
+	dbTemplateLookup := make(map[string]model.TemplateModel)
+	for _, dbTemplate := range dbTemplates {
+		dbTemplateIds = append(dbTemplateIds, dbTemplate.ID)
+		dbTemplateCommits[dbTemplate.Commit] = true
+		dbTemplateLookup[dbTemplate.FolderName] = dbTemplate
+	}
+
+	// Pick out db Version IDs
+	versionQuery := `
+	SELECT *
+	FROM catalog_version
+	WHERE template_id IN (?)`
+	dbVersions := []model.VersionModel{}
+	m.db.Raw(versionQuery, dbTemplateIds).Find(&dbVersions)
+
+	dbVersionCommits := make(map[string]bool)
+	for _, dbVersion := range dbVersions {
+		dbVersionCommits[dbVersion.Commit] = true
+	}
+
+	templateCommits := make(map[string]bool)
+	templateLookup := make(map[string]bool)
+	versionCommits := make(map[string]bool)
+
+	for _, template := range templates {
+		templateCommits[template.Commit] = true
+		templateLookup[template.FolderName] = true
+		for _, version := range template.Versions {
+			versionCommits[version.Commit] = true
+		}
 	}
 
+	tx := m.db.Begin()
+
 	var catalogModel model.CatalogModel
 	catalogModel.Catalog = catalog
 	catalogModel.Commit = newCommit
-	if err := tx.Create(&catalogModel).Error; err != nil {
-		tx.Rollback()
-		return err
-	}
 
-	for _, template := range templates {
-		template.CatalogId = catalogModel.ID
-		template.EnvironmentId = catalogModel.EnvironmentId
-		templateModel := model.TemplateModel{
-			Template: template,
+	// not found in db, create catalog
+	if reflect.DeepEqual(dbCatalog, model.CatalogModel{}) {
```

Comment:
This code isn't very readable or self-evident. Is there a simpler, more obvious test you could do like `dbCatalog.ID == 0` that would more obviously indicate you are just checking to see if no result was found?

Doing so would also be more efficient than a reflect.DeepEqual (though that is admittedly a micro-optimization that would have negligible real impact).